# Import the Packages

In [1]:
import numpy as np

import pandas as pd

import matplotlib.pyplot as plt

import seaborn as sns

# Read the Data

In [2]:
data_path = r"D:\Flight-Price-Prediction\data\flight_price.csv"

flights_df = pd.read_csv(data_path)

In [3]:
flights_df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

**Observations**:
1. The `Airline` column has all the airline name categories.
2. The `Date_of_journey` column has all the entries as dates with the day first format.
3. The `Source` and `Destination` columns have city names and is categorical in nature.
4. The `Route` column marks the route of all the flights with information of the stops included along with it.
5. The `Dep Time` and `Arrival_Time` columns need some cleaning as they appear mixed types with time and date information.
6. The `Duration` column has the flight duration in both hours and minutes and has to be converted in a sinle format for better interpretation.
7. The `Total_Stops` is also categorical in natures and has values of total stops in text and should be converted to integer marking the number of stops the flight takes between its complete route.
8. The `Addtitional_info` column has to be investigated further.
9. The `Price` column is the target column and has price as integers.

In [6]:
print(f'The shape of the flights data has {flights_df.shape[0]} rows and {flights_df.shape[1]} columns')

The shape of the flights data has 10683 rows and 11 columns


In [7]:
flights_df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [5]:
flights_df.sample(20)

Airline Date_of_Journey    Source Destination  \
5847  Jet Airways      27/05/2019     Delhi      Cochin   
9853       IndiGo       6/04/2019  Banglore       Delhi   
5273     Air Asia      12/04/2019  Banglore       Delhi   
5701    Air India      01/03/2019  Banglore   New Delhi   
9511      Vistara      12/03/2019    Mumbai   Hyderabad   
284        IndiGo       6/05/2019   Kolkata    Banglore   
9745  Jet Airways       9/05/2019     Delhi      Cochin   
4419  Jet Airways      18/04/2019  Banglore       Delhi   
3054  Jet Airways      18/06/2019  Banglore       Delhi   
9941     SpiceJet      24/03/2019    Mumbai   Hyderabad   
2447  Jet Airways       6/06/2019     Delhi      Cochin   
9809    Air India      24/06/2019     Delhi      Cochin   
3384       IndiGo      21/05/2019    Mumbai   Hyderabad   
770   Jet Airways       6/05/2019     Delhi      Cochin   
3207       IndiGo      24/05/2019   Kolkata    Banglore   
6408       IndiGo       3/05/2019  Banglore       Delhi   
7133  Jet Airways      24/06/2019     Delhi      Cochin   
5689  Jet Airways      15/04/2019  Banglore       Delhi   
5521       IndiGo      15/03/2019  Banglore   New Delhi   
2963     SpiceJet      27/05/2019   Chennai     Kolkata   

                      Route Dep_Time  Arrival_Time Duration Total_Stops  \
5847        DEL → BOM → COK    22:50  19:00 28 May  20h 10m      1 stop   
9853              BLR → DEL    10:30         13:20   2h 50m    non-stop   
5273              BLR → DEL    23:55  02:45 13 Apr   2h 50m    non-stop   
5701  BLR → BOM → BHO → DEL    08:50  23:25 02 Mar  38h 35m     2 stops   
9511        BOM → DEL → HYD    07:30         19:55  12h 25m      1 stop   
284               CCU → BLR    21:25  00:05 07 May   2h 40m    non-stop   
9745        DEL → BOM → COK    08:00  04:25 10 May  20h 25m      1 stop   
4419              BLR → DEL    17:45         20:45       3h    non-stop   
3054              BLR → DEL    08:20         11:20       3h    non-stop   
9941              BOM → HYD    13:10         14:30   1h 20m    non-stop   
2447        DEL → BOM → COK    20:55  19:00 07 Jun   22h 5m      1 stop   
9809  DEL → JDH → BOM → COK    12:55  19:15 25 Jun  30h 20m     2 stops   
3384              BOM → HYD    16:55         18:25   1h 30m    non-stop   
770         DEL → BOM → COK    13:00  04:25 07 May  15h 25m      1 stop   
3207        CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
6408              BLR → DEL    07:10         10:05   2h 55m    non-stop   
7133  DEL → NAG → BOM → COK    14:35  12:35 25 Jun      22h     2 stops   
5689              BLR → DEL    07:10         10:10       3h    non-stop   
5521              BLR → DEL    22:20  01:10 16 Mar   2h 50m    non-stop   
2963              MAA → CCU    09:45         12:00   2h 15m    non-stop   

                   Additional_Info  Price  
5847                       No info  16079  
9853                       No info   4423  
5273                       No info   7080  
5701                       No info  25430  
9511                       No info  12080  
284                        No info   4804  
9745   In-flight meal not included  12373  
4419                       No info   7229  
3054                       No info   8016  
9941                       No info   2753  
2447                       No info  14714  
9809                       No info  12436  
3384                       No info   3175  
770                        No info  15554  
3207                       No info   6218  
6408                       No info   4823  
7133                       No info  13376  
5689                       No info   7229  
5521                       No info   8855  
2963  No check-in baggage included   3543

In [9]:
# data types of columns in the data

flights_df.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [10]:
# info about the dataframe

flights_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [11]:
# checking for missing values in the data

flights_df.isna().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [27]:
# filter out the rows that have missing values

(
    flights_df
    .loc[
     flights_df
    .isna()
    .any(axis=1)
    , :
    ]
    
)

Airline Date_of_Journey Source Destination Route Dep_Time  \
9039  Air India       6/05/2019  Delhi      Cochin   NaN    09:45   

      Arrival_Time Duration Total_Stops Additional_Info  Price  
9039  09:25 07 May  23h 40m         NaN         No info   7480

Index value `9039` has missing values in columns `Route` and `Total Stops`

In [39]:
# check for duplicate rows in the data

(
    flights_df
    .duplicated()
    .sum()
)

220

**220 rows in the data are duplicated**

In [47]:
(
    flights_df
    .loc[
    flights_df
    .duplicated(keep=False)
    ]
    .sort_values(by=['Date_of_Journey','Airline'])
    
)

Airline Date_of_Journey    Source Destination  \
6321    Air India      01/03/2019  Banglore   New Delhi   
9848    Air India      01/03/2019  Banglore   New Delhi   
572     Air India      03/03/2019  Banglore   New Delhi   
8168    Air India      03/03/2019  Banglore   New Delhi   
1552       IndiGo      03/03/2019  Banglore   New Delhi   
...           ...             ...       ...         ...   
8446  Jet Airways       9/06/2019     Delhi      Cochin   
8496  Jet Airways       9/06/2019     Delhi      Cochin   
8967  Jet Airways       9/06/2019     Delhi      Cochin   
9191  Jet Airways       9/06/2019     Delhi      Cochin   
9216  Jet Airways       9/06/2019     Delhi      Cochin   

                      Route Dep_Time  Arrival_Time Duration Total_Stops  \
6321  BLR → BOM → AMD → DEL    08:50  23:55 02 Mar   39h 5m     2 stops   
9848  BLR → BOM → AMD → DEL    08:50  23:55 02 Mar   39h 5m     2 stops   
572               BLR → DEL    21:10         23:55   2h 45m    non-stop   
8168              BLR → DEL    21:10         23:55   2h 45m    non-stop   
1552              BLR → DEL    16:55         19:45   2h 50m    non-stop   
...                     ...      ...           ...      ...         ...   
8446  DEL → NAG → BOM → COK    06:45  12:35 10 Jun  29h 50m     2 stops   
8496  DEL → JAI → BOM → COK    09:40  12:35 10 Jun  26h 55m     2 stops   
8967  DEL → JAI → BOM → COK    09:40  04:25 10 Jun  18h 45m     2 stops   
9191  DEL → JAI → BOM → COK    09:40  12:35 10 Jun  26h 55m     2 stops   
9216  DEL → IDR → BOM → COK    05:30  04:25 10 Jun  22h 55m     2 stops   

                  Additional_Info  Price  
6321                      No info  17135  
9848                      No info  17135  
572                       No info   7591  
8168                      No info   7591  
1552                      No info   8855  
...                           ...    ...  
8446                      No info  13376  
8496                      No info  13014  
8967  In-flight meal not included  10368  
9191  In-flight meal not included  10368  
9216                      No info  13292  

[415 rows x 11 columns]

In [61]:
(
    flights_df
    .rename(columns=str.lower)
    .columns
)

Index(['airline', 'date_of_journey', 'source', 'destination', 'route',
       'dep_time', 'arrival_time', 'duration', 'total_stops',
       'additional_info', 'price'],
      dtype='object')

## Airline

In [76]:
(
    flights_df['Airline']
    .str.replace('Vistara Premium economy','Vistara')
    .str.replace('Multiple carriers Premium economy','Multiple carriers')
    .str.replace('Jet Airways Business','Jet Airways')
    .str.strip()
    .str.replace(' ','_')
    .str.title()
    .unique()
)

array(['Indigo', 'Air_India', 'Jet_Airways', 'Spicejet',
       'Multiple_Carriers', 'Goair', 'Vistara', 'Air_Asia', 'Trujet'],
      dtype=object)

**Steps**
1. Replace the duplicate flight carriers and make them one.
2. Strip extra whitespaces from categories.
3. Replace the space between two words with "_".
4. Convert the words into Title case.

## Date_of_Journey

In [87]:
(
    pd.to_datetime(flights_df['Date_of_Journey'],dayfirst=True)
)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

**Steps:**
1. Convert the column to datetime.
2. Extract the month and day information out of the datetime column.
3. Drop the original column after the extraction step.

## Source

In [102]:
(
    flights_df['Source']
    .unique()
)

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [97]:
(
    flights_df['Source']
    .value_counts()
)

Source
Delhi       4537
Kolkata     2871
Banglore    2197
Mumbai       697
Chennai      381
Name: count, dtype: int64

**No data cleaning steps to be performed in `Source` column**

## Destination

In [99]:
(
    flights_df['Destination']
    .unique()
)

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

**Steps**
1. Merge the `New Delhi` and `Delhi` category together

In [100]:
(
    flights_df['Destination']
    .value_counts()
)

Destination
Cochin       4537
Banglore     2871
Delhi        1265
New Delhi     932
Hyderabad     697
Kolkata       381
Name: count, dtype: int64

In [105]:
(
    flights_df['Destination']
    .str.replace('New Delhi','Delhi')
    .value_counts()
)

Destination
Cochin       4537
Banglore     2871
Delhi        2197
Hyderabad     697
Kolkata       381
Name: count, dtype: int64

In [101]:
932 + 1265

2197

## Dep_Time

In [119]:
flights_df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [136]:
(
    pd.to_datetime(flights_df['Dep_Time'],format='mixed')
    .dt.time
)

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

**Convert the column to datetime and extract the time information out of it**

# Data Cleaning

In [137]:
def data_cleaning(df:pd.DataFrame):
    return (
        df
        .drop_duplicates()
        .rename(columns=str.lower)
        .assign(
            airline = lambda df_:df_['airline']
                                .str.replace('Vistara Premium economy','Vistara')
                                .str.replace('Multiple carriers Premium economy','Multiple carriers')
                                .str.replace('Jet Airways Business','Jet Airways')
                                .str.strip()
                                .str.replace(' ','_')
                                .str.title(),
            date_of_journey = lambda df_: pd.to_datetime(df_['date_of_journey'],dayfirst=True),
            day_of_journey = lambda df_: df_['date_of_journey'].dt.day,
            month_of_journey = lambda df_: df_['date_of_journey'].dt.month
        )
        .drop(columns='date_of_journey')
        .assign(
            destination = lambda df_: df_['destination']
                                    .str.replace('New Delhi','Delhi')
        )
        .drop(columns=['route'])
        .rename(columns={
            "dep_time": "departure_time"
        })
        .assign(
            departure_time = lambda df_: pd.to_datetime(df_['departure_time'],format='mixed')
                                        .dt.time

        )
    )

In [138]:
data_cleaning(flights_df)

airline    source destination departure_time  arrival_time  \
0           Indigo  Banglore       Delhi       22:20:00  01:10 22 Mar   
1        Air_India   Kolkata    Banglore       05:50:00         13:15   
2      Jet_Airways     Delhi      Cochin       09:25:00  04:25 10 Jun   
3           Indigo   Kolkata    Banglore       18:05:00         23:30   
4           Indigo  Banglore       Delhi       16:50:00         21:35   
...            ...       ...         ...            ...           ...   
10678     Air_Asia   Kolkata    Banglore       19:55:00         22:25   
10679    Air_India   Kolkata    Banglore       20:45:00         23:20   
10680  Jet_Airways  Banglore       Delhi       08:20:00         11:20   
10681      Vistara  Banglore       Delhi       11:30:00         14:10   
10682    Air_India     Delhi      Cochin       10:55:00         19:15   

      duration total_stops additional_info  price  day_of_journey  \
0       2h 50m    non-stop         No info   3897              24   
1       7h 25m     2 stops         No info   7662               1   
2          19h     2 stops         No info  13882               9   
3       5h 25m      1 stop         No info   6218              12   
4       4h 45m      1 stop         No info  13302               1   
...        ...         ...             ...    ...             ...   
10678   2h 30m    non-stop         No info   4107               9   
10679   2h 35m    non-stop         No info   4145              27   
10680       3h    non-stop         No info   7229              27   
10681   2h 40m    non-stop         No info  12648               1   
10682   8h 20m     2 stops         No info  11753               9   

       month_of_journey  
0                     3  
1                     5  
2                     6  
3                     5  
4                     3  
...                 ...  
10678                 4  
10679                 4  
10680                 4  
10681                 3  
10682                 5  

[10463 rows x 11 columns]